cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,2423 @@ func TestCreateVPCSubnet(t *testing.T) {
+	t.Run("When it matches SecurityGroupRule of protocolType SecurityGroupRuleSecurityGroupRuleProtocolIcmp", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		remote := infrav1beta2.VPCSecurityGroupRuleRemote{
+			SecurityGroupName: ptr.To("securityGroupName"),
+			RemoteType:        infrav1beta2.VPCSecurityGroupRuleRemoteTypeSG,
+		}
+		rules := infrav1beta2.VPCSecurityGroupRule{
+			Direction: infrav1beta2.VPCSecurityGroupRuleDirectionOutbound,
+			Destination: &infrav1beta2.VPCSecurityGroupRulePrototype{
+				Remotes:  append([]infrav1beta2.VPCSecurityGroupRuleRemote{}, remote),
+				Protocol: infrav1beta2.VPCSecurityGroupRuleProtocolTCP,
+				PortRange: &infrav1beta2.VPCSecurityGroupPortRange{MaximumPort: 65535,
+					MinimumPort: 1,
+				},
+				ICMPCode: (ptr.To(int64(12))),
+				ICMPType: (ptr.To(int64(3))),
+			},
+		}
+		vpcSecurityGroupRule := vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolIcmp{
+			Direction: ptr.To("outbound"),
+			Protocol:  ptr.To("tcp"),
+			Remote: &vpcv1.SecurityGroupRuleRemote{
+				CRN: ptr.To("crn"),
+			},
+			ID:   ptr.To("ruleID"),
+			Code: (ptr.To(int64(12))),
+			Type: (ptr.To(int64(3))),
+		}
+
+		vpcSecurityGroupRules := append([]vpcv1.SecurityGroupRuleIntf{}, &vpcSecurityGroupRule)
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			ID:    ptr.To("securityGroupID"),
+			Rules: append([]*infrav1beta2.VPCSecurityGroupRule{}, &rules),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroupByName(gomock.Any()).Return(&vpcv1.SecurityGroup{Name: ptr.To("crn"), CRN: ptr.To("crn")}, nil)
+		ruleID, match, err := clusterScope.validateSecurityGroupRule(vpcSecurityGroupRules, rules.Direction, rules.Destination, remote)
+		g.Expect(*ruleID).To(BeEquivalentTo("ruleID"))
+		g.Expect(match).To(BeTrue())
+		g.Expect(err).To(BeNil())
+	})
+	t.Run("When it doesn't match SecurityGroupRule of protocolType SecurityGroupRuleSecurityGroupRuleProtocolIcmp", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		remote := infrav1beta2.VPCSecurityGroupRuleRemote{
+			SecurityGroupName: ptr.To("securityGroupName"),
+			RemoteType:        infrav1beta2.VPCSecurityGroupRuleRemoteTypeSG,
+		}
+		rules := infrav1beta2.VPCSecurityGroupRule{
+			Direction: infrav1beta2.VPCSecurityGroupRuleDirectionOutbound,
+			Destination: &infrav1beta2.VPCSecurityGroupRulePrototype{
+				Remotes:  append([]infrav1beta2.VPCSecurityGroupRuleRemote{}, remote),
+				Protocol: infrav1beta2.VPCSecurityGroupRuleProtocolTCP,
+				PortRange: &infrav1beta2.VPCSecurityGroupPortRange{MaximumPort: 65535,
```

Comment:
Port range is not valid for icmp right, can you please remove here and where ever not needed?

Reply:
It is needed in every case, we are processing port range before checking the SecurityGroupRule  type
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_cluster.go#L1514

Reply:
But we won't access the value in icmp right and since it's not a pointer [default value](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/3ee7b716c11e7989a84c833d492b5ef642479e85/api/v1beta2/types.go#L422) would be there, no need to define like this. Same as the case for protocol all.

Reply:
Oh in that case, can we move the [port range](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_cluster.go#L1514toL1515) define section inside the icmp?